src/Integration.Vsix/SonarLintTracker/IssueTracker.cs

```diff
@@ -0,0 +1,204 @@
+    /// </remarks>
+    internal class IssueTracker : ITagger<IErrorTag>, IDisposable
+    {
+        private readonly TaggerProvider provider;
+        private readonly ITextBuffer textBuffer;
+
+        private ITextSnapshot currentSnapshot;
+        private NormalizedSnapshotSpanCollection dirtySpans;
+
+        internal string FilePath { get; private set; }
+        internal string Charset { get; }
+        internal SnapshotFactory Factory { get; }
+
+        internal IssuesSnapshot Snapshot { get; set; }
+
+        internal IssueTracker(TaggerProvider provider, ITextBuffer buffer, ITextDocument document)
+        {
+            this.provider = provider;
+            this.textBuffer = buffer;
+            this.currentSnapshot = buffer.CurrentSnapshot;
+            this.dirtySpans = new NormalizedSnapshotSpanCollection();
+
+            this.FilePath = document.FilePath;
+            this.Charset = document.Encoding.WebName;
+            this.Factory = new SnapshotFactory(new IssuesSnapshot(this.FilePath, 0, new List<IssueMarker>()));
+
+            document.FileActionOccurred += FileActionOccurred;
+
+            this.Initialize();
+        }
+
+        private void FileActionOccurred(object sender, TextDocumentFileActionEventArgs e)
+        {
+            if (e.FileActionType == FileActionTypes.DocumentRenamed)
+            {
+                provider.Rename(FilePath, e.FilePath);
+                FilePath = e.FilePath;
+            }
+            else if (e.FileActionType == FileActionTypes.ContentSavedToDisk)
+            {
+                provider.RequestAnalysis(FilePath, Charset);
+            }
+        }
+
+        private void Initialize()
+        {
+            textBuffer.ChangedLowPriority += this.OnBufferChange;
+            provider.AddIssueTracker(this);
+            provider.RequestAnalysis(FilePath, Charset);
+        }
+
+        public void Dispose()
+        {
+            textBuffer.ChangedLowPriority -= this.OnBufferChange;
+            provider.RemoveIssueTracker(this);
+        }
+
+        private void OnBufferChange(object sender, TextContentChangedEventArgs e)
+        {
+            UpdateDirtySpans(e);
+
+            var newMarkers = TranslateMarkerSpans();
+
+            SnapToNewSnapshot(newMarkers);
+        }
+
+        private void UpdateDirtySpans(TextContentChangedEventArgs e)
+        {
+            currentSnapshot = e.After;
+
+            var newDirtySpans = dirtySpans.CloneAndTrackTo(e.After, SpanTrackingMode.EdgeInclusive);
+
+            foreach (var change in e.Changes)
+            {
+                newDirtySpans = NormalizedSnapshotSpanCollection.Union(newDirtySpans, new NormalizedSnapshotSpanCollection(e.After, change.NewSpan));
+            }
+
+            dirtySpans = newDirtySpans;
+        }
+
+        private IssuesSnapshot TranslateMarkerSpans()
+        {
+            var oldSnapshot = this.Factory.CurrentSnapshot;
+            var newMarkers = oldSnapshot.IssueMarkers
```

Comment:
With the change to the `IssuesSnapshot` ctor you no longer need the `.ToList()`.